<a href="https://colab.research.google.com/github/Neyker8a/challenge2-data-science-LATAM/blob/main/TelecomX_LATAM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#📌 Extracción

In [149]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

url = "https://raw.githubusercontent.com/ingridcristh/challenge2-data-science-LATAM/main/TelecomX_Data.json"

df = pd.read_json(url)
df.head()

,customerID,Churn,customer,phone,internet,account
0,0002-ORFBO,No,"{'gender': 'Female', 'SeniorCitizen': 0, 'Part...","{'PhoneService': 'Yes', 'MultipleLines': 'No'}","{'InternetService': 'DSL', 'OnlineSecurity': '...","{'Contract': 'One year', 'PaperlessBilling': '..."
1,0003-MKNFE,No,"{'gender': 'Male', 'SeniorCitizen': 0, 'Partne...","{'PhoneService': 'Yes', 'MultipleLines': 'Yes'}","{'InternetService': 'DSL', 'OnlineSecurity': '...","{'Contract': 'Month-to-month', 'PaperlessBilli..."
2,0004-TLHLJ,Yes,"{'gender': 'Male', 'SeniorCitizen': 0, 'Partne...","{'PhoneService': 'Yes', 'MultipleLines': 'No'}","{'InternetService': 'Fiber optic', 'OnlineSecu...","{'Contract': 'Month-to-month', 'PaperlessBilli..."
3,0011-IGKFF,Yes,"{'gender': 'Male', 'SeniorCitizen': 1, 'Partne...","{'PhoneService': 'Yes', 'MultipleLines': 'No'}","{'InternetService': 'Fiber optic', 'OnlineSecu...","{'Contract': 'Month-to-month', 'PaperlessBilli..."
4,0013-EXCHZ,Yes,"{'gender': 'Female', 'SeniorCitizen': 1, 'Part...","{'PhoneService': 'Yes', 'MultipleLines': 'No'}","{'InternetService': 'Fiber optic', 'OnlineSecu...","{'Contract': 'Month-to-month', 'PaperlessBilli..."


In [150]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


#🔧 Transformación

In [151]:
df.isnull().sum()

,0
customerID,0
Churn,0
customer,0
phone,0
internet,0
account,0


In [152]:

# Aplanar la columna 'customer'
df_customer = pd.json_normalize(df['customer'])
df_customer.columns = ['customer_' + col for col in df_customer.columns] # Prefijo para evitar colisiones

# Aplanar la columna 'phone'
df_phone = pd.json_normalize(df['phone'])
df_phone.columns = ['phone_' + col for col in df_phone.columns]

# Aplanar la columna 'internet'
df_internet = pd.json_normalize(df['internet'])
df_internet.columns = ['internet_' + col for col in df_internet.columns]

# Aplanar la columna 'account'
df_account = pd.json_normalize(df['account'])
df_account.columns = ['account_' + col for col in df_account.columns]

# Unir las nuevas columnas a nuestro DataFrame original y eliminar las columnas de diccionario
df_flattened = pd.concat([
    df.drop(columns=['customer', 'phone', 'internet', 'account']),
    df_customer,
    df_phone,
    df_internet,
    df_account
], axis=1)

print("DataFrame original y sus columnas de diccionario aplanadas:")
display(df_flattened.head())

DataFrame original y sus columnas de diccionario aplanadas:


,customerID,Churn,customer_gender,customer_SeniorCitizen,customer_Partner,customer_Dependents,customer_tenure,phone_PhoneService,phone_MultipleLines,internet_InternetService,...,internet_OnlineBackup,internet_DeviceProtection,internet_TechSupport,internet_StreamingTV,internet_StreamingMovies,account_Contract,account_PaperlessBilling,account_PaymentMethod,account_Charges.Monthly,account_Charges.Total
0,0002-ORFBO,No,Female,0,Yes,Yes,9,Yes,No,DSL,...,Yes,No,Yes,Yes,No,One year,Yes,Mailed check,65.6,593.3
1,0003-MKNFE,No,Male,0,No,No,9,Yes,Yes,DSL,...,No,No,No,No,Yes,Month-to-month,No,Mailed check,59.9,542.4
2,0004-TLHLJ,Yes,Male,0,No,No,4,Yes,No,Fiber optic,...,No,Yes,No,No,No,Month-to-month,Yes,Electronic check,73.9,280.85
3,0011-IGKFF,Yes,Male,1,Yes,No,13,Yes,No,Fiber optic,...,Yes,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,98.0,1237.85
4,0013-EXCHZ,Yes,Female,1,Yes,No,3,Yes,No,Fiber optic,...,No,No,Yes,Yes,No,Month-to-month,Yes,Mailed check,83.9,267.4


In [153]:
df_flattened.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer_gender            7267 non-null   object 
 3   customer_SeniorCitizen     7267 non-null   int64  
 4   customer_Partner           7267 non-null   object 
 5   customer_Dependents        7267 non-null   object 
 6   customer_tenure            7267 non-null   int64  
 7   phone_PhoneService         7267 non-null   object 
 8   phone_MultipleLines        7267 non-null   object 
 9   internet_InternetService   7267 non-null   object 
 10  internet_OnlineSecurity    7267 non-null   object 
 11  internet_OnlineBackup      7267 non-null   object 
 12  internet_DeviceProtection  7267 non-null   object 
 13  internet_TechSupport       7267 non-null   objec

In [154]:
# Verificar la cantidad de filas duplicadas en el DataFrame aplanado
num_duplicates_flattened = df_flattened.duplicated().sum()
print(f"Número de filas duplicadas en el DataFrame aplanado: {num_duplicates_flattened}")


Número de filas duplicadas en el DataFrame aplanado: 0


In [155]:
# Normalizar todas las columnas de tipo 'object' a minúsculas
for col in df_flattened.select_dtypes(include='object').columns:
    # Convertir a string y luego a minúsculas
    df_flattened[col] = df_flattened[col].astype(str).str.lower()

print("DataFrame con columnas de texto normalizadas a minúsculas:")
df_flattened.head()
df_flattened.info()


DataFrame con columnas de texto normalizadas a minúsculas:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer_gender            7267 non-null   object 
 3   customer_SeniorCitizen     7267 non-null   int64  
 4   customer_Partner           7267 non-null   object 
 5   customer_Dependents        7267 non-null   object 
 6   customer_tenure            7267 non-null   int64  
 7   phone_PhoneService         7267 non-null   object 
 8   phone_MultipleLines        7267 non-null   object 
 9   internet_InternetService   7267 non-null   object 
 10  internet_OnlineSecurity    7267 non-null   object 
 11  internet_OnlineBackup      7267 non-null   object 
 12  internet_DeviceProtection  7267 non-null   ob

In [156]:
df_flattened['account_Charges.Total'].dtype
df_flattened["account_Charges.Total"] = pd.to_numeric(df_flattened["account_Charges.Total"], errors="coerce")
df_flattened.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer_gender            7267 non-null   object 
 3   customer_SeniorCitizen     7267 non-null   int64  
 4   customer_Partner           7267 non-null   object 
 5   customer_Dependents        7267 non-null   object 
 6   customer_tenure            7267 non-null   int64  
 7   phone_PhoneService         7267 non-null   object 
 8   phone_MultipleLines        7267 non-null   object 
 9   internet_InternetService   7267 non-null   object 
 10  internet_OnlineSecurity    7267 non-null   object 
 11  internet_OnlineBackup      7267 non-null   object 
 12  internet_DeviceProtection  7267 non-null   object 
 13  internet_TechSupport       7267 non-null   objec

In [157]:
df_flattened["account_Charges.Total"].isnull().sum()

np.int64(11)

In [158]:
df_flattened = df_flattened.dropna(subset=["account_Charges.Total"])

In [159]:
df_flattened["account_Charges.Total"].isnull().sum()

np.int64(0)

In [160]:
df_flattened.duplicated().sum()

np.int64(0)

In [161]:
df_flattened.loc[:, "cuentas_diarias"] = df_flattened["account_Charges.Monthly"] / 30
df_flattened[["account_Charges.Monthly", "cuentas_diarias"]].head()

,account_Charges.Monthly,cuentas_diarias
0,65.6,2.186667
1,59.9,1.996667
2,73.9,2.463333
3,98.0,3.266667
4,83.9,2.796667


In [162]:
binary_cols = [
    col for col in df_flattened.columns
    if set(df_flattened[col].dropna().unique()).issubset({"yes", "no"})
]


In [163]:
df_flattened.columns = (
    df_flattened.columns
    .str.replace(".", "_", regex=False)
    .str.lower()
)

df_flattened.head()
df_flattened.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7256 entries, 0 to 7266
Data columns (total 22 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerid                 7256 non-null   object 
 1   churn                      7256 non-null   object 
 2   customer_gender            7256 non-null   object 
 3   customer_seniorcitizen     7256 non-null   int64  
 4   customer_partner           7256 non-null   object 
 5   customer_dependents        7256 non-null   object 
 6   customer_tenure            7256 non-null   int64  
 7   phone_phoneservice         7256 non-null   object 
 8   phone_multiplelines        7256 non-null   object 
 9   internet_internetservice   7256 non-null   object 
 10  internet_onlinesecurity    7256 non-null   object 
 11  internet_onlinebackup      7256 non-null   object 
 12  internet_deviceprotection  7256 non-null   object 
 13  internet_techsupport       7256 non-null   object 
 1

In [164]:
df_flattened["churn"].unique()
(df_flattened["churn"] == "").sum()
df_flattened["churn"] = df_flattened["churn"].str.strip().str.lower()
df_flattened = df_flattened[df_flattened["churn"] != ""]
(df_flattened["churn"] == "").sum()


np.int64(0)

In [165]:
df_flattened["churn"] = df_flattened["churn"].map({
    "yes": 1,
    "no": 0
})
df_flattened["churn"].unique()


array([0, 1])

In [166]:
df_flattened.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7266
Data columns (total 22 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerid                 7032 non-null   object 
 1   churn                      7032 non-null   int64  
 2   customer_gender            7032 non-null   object 
 3   customer_seniorcitizen     7032 non-null   int64  
 4   customer_partner           7032 non-null   object 
 5   customer_dependents        7032 non-null   object 
 6   customer_tenure            7032 non-null   int64  
 7   phone_phoneservice         7032 non-null   object 
 8   phone_multiplelines        7032 non-null   object 
 9   internet_internetservice   7032 non-null   object 
 10  internet_onlinesecurity    7032 non-null   object 
 11  internet_onlinebackup      7032 non-null   object 
 12  internet_deviceprotection  7032 non-null   object 
 13  internet_techsupport       7032 non-null   object 
 1

#📊 Carga y análisis

In [167]:
df_flattened.describe()

,churn,customer_seniorcitizen,customer_tenure,account_charges_monthly,account_charges_total,cuentas_diarias
count,7032.000000,7032.000000,7032.000000,7032.000000,7032.000000,7032.000000
mean,0.265785,0.162400,32.421786,64.798208,2283.300441,2.159940
std,0.441782,0.368844,24.545260,30.085974,2266.771362,1.002866
min,0.000000,0.000000,1.000000,18.250000,18.800000,0.608333
25%,0.000000,0.000000,9.000000,35.587500,401.450000,1.186250
50%,0.000000,0.000000,29.000000,70.350000,1397.475000,2.345000
75%,1.000000,0.000000,55.000000,89.862500,3794.737500,2.995417
max,1.000000,1.000000,72.000000,118.750000,8684.800000,3.958333


Interpretación: churn

Media = 0.265 → 26.5% de los clientes cancelan el servicio

73.5% permanecen

Esto indica una clase desbalanceada moderada, algo común en problemas de churn.

Interpretación: customer_seniorcityzen

Aproximadamente 16.2% de los clientes son adultos mayores.

La mayoría de clientes no pertenecen a este grupo.

Interpretación: tenure

Los clientes tienen en promedio 32 meses de permanencia.

La mitad de los clientes tiene menos de 29 meses.

Algunos clientes llevan hasta 6 años con la empresa.

Esto sugiere una mezcla de clientes nuevos y clientes leales.

Interpretación: account_charges_monthly

El cargo mensual promedio es ~65.

Algunos clientes pagan menos de 20 y otros más de 118.

La mediana es mayor que el promedio, lo que sugiere ligera asimetría en la distribución.

Interpretación: cuentas_diarias

El gasto diario promedio es 2.16 unidades monetarias.

Existe variabilidad entre clientes con planes más económicos y más costosos.

#📄Informe final